### PR TITLE
[blog] Add post about writing software for core networks

### PR DESCRIPTION
**Pull request by david-wg2**

### CODE REVIEW

**Comment by Waldumwg2** [comment]
Good post! Let's find a way to make sure we track new signups and have Google Analytics ready before posting it

File: pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md [hunk 1]
```diff
@@ -0,0 +1,157 @@
...
+        logger.warn("Unable to send MMS to ${phone.e164}: ${response.description} (${response.status.name})")
+        false
+    }
+} catch (e: Exception) {
+    logger.warn("Unable to send MMS to ${phone.e164}: ${e.message}")
+    false
+}
+```
+
+This is also pretty straightforward, except for the part where the sender is also the receiver. In the future this will be a bit less confusing, since we’ll be sending the MMS from the product (“VoiceBox”) instead of the subscriber itself.
```
Comment (jorunfa) [inline pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md:127]: *Personal opinions which might not reflect other people's opinions inc*

*Feedback 1*
I felt proud when I started reading the article and found the story being told interesting (you should probably fire me if I didn't).

*Feedback 2*
I read the article while I was out walking the dog. Somewhere around this bit I got a bit weary of reading the article and felt like taking a break from reading it. It might be because I'm already familiar with the content, but it felt a bit wordy to me, repeating a bit of what's written in code in plain english. This feels appropriate when writing a tutorial but felt verbose to me in an article geared at showcasing/telling a story about tech.

I'll try to highlight some of the sentences I found superfluous.

*Feedback 3*
Although not perfect I also think it would be great if we could a screenshot of a received mms with an audiofile or whatever, so you get the full story/feature. Probably at the beginning there with the first couple of pictures.

File: pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md [hunk 2]
```diff
@@ -0,0 +1,157 @@
...
+1. The subscriber signs in and enables the desired functionality in VoiceBox
+2. The next time the subscriber receives a voicemail, an event is fired by Working Group Two’s core network, which VoiceBox receives
+3. VoiceBox triggers an SMS/MMS send using our APIs
+4. The subscriber receives an SMS/MMS from the sender “VoiceBox”
+
+In a few months time this ideal world should be reality, but at the time of writing (mid August 2020) we’re missing the events API and the “Send from Product” API. Currently VoiceBox works like this (changes are highlighted):
+1. The subscriber signs in and enables the desired functionality in VoiceBox
+2. The next time the subscriber receives a voicemail, **VoiceBox will discover it by polling**
+3. VoiceBox triggers an MMS send using our APIs
+4. The subscriber receives an **MMS from their own number**
```
Comment (jorunfa) [inline pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md:47]: The instances in these lists are full sentences; as far as I know they should be punctuated then.
(first article I found https://www.businesswritingblog.com/business_writing/2012/01/punctuating-bullet-points.html)

File: pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md [hunk 3]
```diff
@@ -0,0 +1,157 @@
...
+
+It’s not perfect, but it still demonstrates the potential of the platform. None of this can happen without the subscriber’s consent though, so in the next section we’ll have a look at how that works.
+
+## Obtaining user consent
+
+The app we are building is touching sensitive data, and there is no way we can do that without asking the subscriber if it’s okay. The Working Group Two platform includes an oauth implementation with pin roundtrip authentication, which means that we can be sure that the subscriber (or someone in control of the subscription) has consented to the subscription being controlled by a third party.
```
Comment (jorunfa) [inline pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md:53]:
```suggestion
The app we are building is touching sensitive data, and there is no way we can do that without asking the subscriber if it’s okay. The Working Group Two platform includes an oauth implementation with pin roundtrip authentication, which means that we can be reasonably sure that the subscriber has consented.
```

File: pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md [hunk 4]
```diff
@@ -0,0 +1,157 @@
...
+
+The login is branded to look like the product (notice the pink action button), while the consent screen is branded to look like the operator (in this case, the Swedish operator <a href="https://vimla.se" target="_blank">Vimla</a>, which uses our platform):
+
+<div class="post-images threes">
+    <img src="/img/blog/building-software-for-a-telecom-core-network/msisdn-screen.png" alt="ID login screen">
+    <img src="/img/blog/building-software-for-a-telecom-core-network/pin-screen.png" alt="ID pin screen">
+    <img src="/img/blog/building-software-for-a-telecom-core-network/consent-screen.png" alt="ID consent screen">
+</div>
+
+This follows a standard <a href="https://oauth.net/2/" target="_blank">OAuth</a> flow, which means that when the subscriber taps “Accept”, Working Group Two redirects the subscriber to the third party. This redirect includes a code as a query parameter, which the third party has to use to obtain an access token and a refresh token. This access token gives the third party access to act on behalf of the subscriber. In this case it will let the third party fetch Voicemails and send MMS from the subscriber, so it’s very important to keep it safe.
```
Comment (jorunfa) [inline pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md:65]: It's probably fine to use `oauth` colloquially elsewhere, but then maybe call it `OAuth 2` here where you reference it properly.

```suggestion
This follows a standard <a href="https://oauth.net/2/" target="_blank">OAuth 2</a> flow, which means that when the subscriber taps “Accept”, Working Group Two redirects the subscriber to the third party. This redirect includes a code as a query parameter, which the third party has to use to obtain an access token and a refresh token. This access token gives the third party access to act on behalf of the subscriber. In this case it will let the third party fetch Voicemails and send MMS from the subscriber, so it’s very important to keep it safe.
```

Comment (jorunfa) [inline pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md:65]:
```suggestion
This follows a standard <a href="https://oauth.net/2/" target="_blank">OAuth</a> flow, which means that when the subscriber taps “Accept”, Working Group Two redirects the subscriber to the third party, granting them access to act on the users behalf.
```

If this is an "here is our tech" and not a _tutorial_ then I'd rather do without the mini-OAuth 2 explanation.

File: pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md [hunk 5]
```diff
@@ -0,0 +1,157 @@
...
+    } else {
+        logger.warn("No voicemail metadata for $uuid")
+        return null
+    }
+}
+```
+
+The first line of the function declares a `GetVoicemailRequest`, which is a class from the official Java client (this example is written in Kotlin, but Kotlin and Java are interoperable languages). Inside the `try`, we pass the OAuth credentials and the request to a `blockingStub`, which is of the type `VoicemailMediaServiceGrpc.VoicemailMediaServiceBlockingStub` (another class we get from the client). If the voicemail metadata is present, we have everything we need to return a VoiceBox specific `Voicemail` class, if not we have to log and return null. Unlike REST APIs, we don't have to think about writing client code and handling HTTP responses, this is taken care of by the client library.
+
+As we mentioned earlier, once we have the `Voicemail`, we send it to the subscriber using their own number as the sender:
```
Comment (jorunfa) [inline pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md:98]:
```suggestion
We use the official library to fetch a voicemail based on the UUID of the voicemail we've fetched at an earlier point.
Unlike REST APIs, we don't have to think about writing client code and handling HTTP responses, this is taken care of by the client library.

Once we have the `voicemail`, we send it to the subscriber using their own number as the sender:
```

File: pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md [hunk 6]
```diff
@@ -0,0 +1,157 @@
...
+One of the most common complaints people have with their Voicemail service in Europe is that you have to call in to your Voicemail to hear your messages. In order to play the message you have to navigate through a slow voice menu using your dialpad. We believed this would be easy to improve.
+
+Our product idea was simple. When Alice leaves a voicemail message for Bob, Bob receives this message either as an audio file or as a speech-to-text transcript.
+
+<div class="post-images">
+    <img src="/img/blog/building-software-for-a-telecom-core-network/voicebox-splash.png" alt="VoiceBox splash screen">
+    <img src="/img/blog/building-software-for-a-telecom-core-network/voicebox-home.png" alt="VoiceBox home screen">
+</div>
+
+The product idea isn’t what most people would call revolutionary, but almost all of the world's mobile operators have hermetically sealed core networks. This means it would be impossible to build this product without lawyering up and coming to some sort of agreement with one of them. The Working Group Two core network, however, is open and provides API access to Voicemail, MMS and SMS  (among other things), which is just what we need to build our product.
```
Comment (jorunfa) [inline pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md:23]: Could also mention something along the lines of that there are millions/billions of people who have voicemail on their subscription, and if we can make a pretty decent improvement on how they receive them, then that is a lot of time saved in total. Don't know if that makes sense or is worth mentioning, but I thought I'd mention it.

File: pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md (hunk 2 above)
Comment (david-wg2) [reply pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md:47]: Fixed (locally).

File: pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md (hunk 3 above)
Comment (david-wg2) [reply pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md:53]: Fixed (locally).

File: pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md (hunk 4 above)
Comment (david-wg2) [reply pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md:65]: Fixed (locally)

Comment (david-wg2) [reply pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md:65]: I'll try to shorten it at least.

Comment (david-wg2) [reply pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md:65]:
```
When the subscriber taps “Accept”, Working Group Two redirects the subscriber to the third party, which receives an "access token" that allows them to act on behalf of the subscriber. In this case it will let the third party fetch Voicemails and send MMS from the subscriber, so it’s important that the access token is kept safe.
```

Better?

File: pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md (hunk 5 above)
Comment (david-wg2) [reply pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md:98]: Fixed (locally)

File: pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md (hunk 6 above)
Comment (david-wg2) [reply pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md:23]: I feel like it's a bit of a digression here, but maybe in the conclusion?

File: pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md (hunk 1 above)
Comment (david-wg2) [reply pages/blog/_posts/2020-08-20-building-software-for-a-telecom-core-network.md:127]: Since I ported my number to Wotel, I can no longer use this feature, so I can't take a screenshot 😅
